wrapper/src/main/java/software/amazon/jdbc/plugin/failover2/FailoverConnectionPlugin.java

```diff
@@ -0,0 +1,619 @@
+  }
+
+  protected <E extends Exception> void dealWithOriginalException(
+      final Throwable originalException,
+      final Throwable wrapperException,
+      final Class<E> exceptionClass) throws E {
+
+    Throwable exceptionToThrow = wrapperException;
+    if (originalException != null) {
+      LOGGER.finer(() -> Messages.get("Failover.detectedException", new Object[]{originalException.getMessage()}));
+      if (this.lastExceptionDealtWith != originalException
+          && shouldExceptionTriggerConnectionSwitch(originalException)) {
+        this.invalidateCurrentConnection();
+        this.pluginService.setAvailability(
+            this.pluginService.getCurrentHostSpec().getAliases(), HostAvailability.NOT_AVAILABLE);
+        try {
+          this.pickNewConnection();
+        } catch (final SQLException e) {
+          throw WrapperUtils.wrapExceptionIfNeeded(exceptionClass, e);
+        }
+        this.lastExceptionDealtWith = originalException;
+      }
+
+      if (originalException instanceof Error) {
+        throw (Error) originalException;
+      }
+      exceptionToThrow = originalException;
+    }
+
+    if (exceptionToThrow instanceof Error) {
+      throw (Error) exceptionToThrow;
+    }
+
+    throw WrapperUtils.wrapExceptionIfNeeded(exceptionClass, exceptionToThrow);
+  }
+
+  protected <E extends Exception> void dealWithIllegalStateException(
+      final IllegalStateException e,
+      final Class<E> exceptionClass) throws E {
+    dealWithOriginalException(e.getCause(), e, exceptionClass);
+  }
+
+  /**
+   * Initiates the failover procedure. This process tries to establish a new connection to an
+   * instance in the topology.
+   *
+   * @param failedHost The host with network errors.
+   * @throws SQLException if an error occurs
+   */
+  protected void failover(final HostSpec failedHost) throws SQLException {
+    this.pluginService.setAvailability(failedHost.asAliases(), HostAvailability.NOT_AVAILABLE);
+
+    if (this.failoverMode == FailoverMode.STRICT_WRITER) {
+      failoverWriter();
+    } else {
+      failoverReader();
+    }
+
+    if (isInTransaction || this.pluginService.isInTransaction()) {
+      if (this.pluginManagerService != null) {
+        this.pluginManagerService.setInTransaction(false);
+      }
+      // "Transaction resolution unknown. Please re-configure session state if required and try
+      // restarting transaction."
+      final String errorMessage = Messages.get("Failover.transactionResolutionUnknownError");
+      LOGGER.info(errorMessage);
+      throw new TransactionStateUnknownSQLException();
+    } else {
+      // "The active SQL connection has changed due to a connection failure. Please re-configure
+      // session state if required. "
+      LOGGER.severe(() -> Messages.get("Failover.connectionChangedError"));
+      throw new FailoverSuccessSQLException();
+    }
+  }
+
+  protected void failoverReader() throws SQLException {
```

Comment:
This method doesn't seem to obey the max failover timeout setting the user has set, it could take longer. I think we need to limit the time spent in this method to obey the setting

Reply:
Good catch. I completely overseen that. I added a simple timeout check condition. However failover time may vary slightly (caused by connect timeout).